### PR TITLE
[shardformer] Support the Command-R model

### PR DESCRIPTION
**Pull request by GuangyaoZhang**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

- [x] https://github.com/hpcaitech/ColossalAI/issues/5831
## 📝 What does this PR do?

Support Command-R model. Most code copied from Llama

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/command.py [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+                past_key_values=outputs.past_key_values,
+                hidden_states=outputs.hidden_states,
+                attentions=outputs.attentions,
+            )
+        else:
+            hidden_states = outputs.get("hidden_states")
+            return {"hidden_states": hidden_states}
+
+
+def get_command_flash_attention_forward(shard_config, sp_mode, sp_group, sp_size):
```
Comment (ver217) [inline colossalai/shardformer/modeling/command.py:337]: Can we merge this and `get_command_seq_parallel_attention_forward` (line 650) into one function?

File: colossalai/shardformer/modeling/command.py [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+        if sp_mode == "all_to_all":
+            attn_output = all_to_all_comm(attn_output, sp_group, scatter_dim=1, gather_dim=2)
+        attn_output = self.o_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+    return forward
+
+
+def get_command_model_forward_for_flash_attn(shard_config: ShardConfig):
```
Comment (ver217) [inline colossalai/shardformer/modeling/command.py:411]: Can we merge this and `get_command_seq_parallel_model_forward` (line 729) into one function?

File: colossalai/shardformer/modeling/command.py (hunk 1 above)
Comment (GuangyaoZhang) [reply colossalai/shardformer/modeling/command.py]: Sure. Merged

File: colossalai/shardformer/policies/command.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+            self.shard_config.enable_sequence_overlap = False
+            self.shard_config.sequence_parallelism_mode = None
+            warnings.warn(
+                f"For Command, sequence parallelism is currently not compatible with pipeline parallelism, set to be False"
+            )
+        sp_mode = self.shard_config.sequence_parallelism_mode if self.shard_config.enable_sequence_parallelism else None
+        sp_size = self.shard_config.sequence_parallel_size if self.shard_config.enable_sequence_parallelism else None
+        sp_group = (
+            self.shard_config.sequence_parallel_process_group if self.shard_config.enable_sequence_parallelism else None
+        )
```
Comment (ver217) [inline colossalai/shardformer/policies/command.py:80]: Can be simplified as `xxx or None`

Comment (GuangyaoZhang) [reply colossalai/shardformer/policies/command.py]: Sure. And I changed all 'xxx if xxx else None' to 'xxx or None' in this Repo

File: colossalai/shardformer/policies/command.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+        self.append_or_create_submodule_replacement(
+            description=SubModuleReplacementDescription(
+                suffix="norm",
+                target_module=norm_cls,
+                kwargs={"sp_partial_derived": sp_partial_derived},
+            ),
+            policy=policy,
+            target_key=CohereModel,
+        )
```
Comment (ver217) [inline colossalai/shardformer/policies/command.py:204]: Do you miss `if` statement?

**Review by ver217 — APPROVED** [review]

**Pull request merged**